### PR TITLE
[Fixes #1572] Improve the extensibility of the default identity UI

### LINKED ISSUE #1572: Improve the extensibility of the default identity UI
Support changing the user type from IdentityUser to TUser while using the default UI.

### PR DESCRIPTION
**Pull request by javiercn**
Allows changing the user type while still reusing the precompiled views.
* Splits page models in two classes:
  * The base class is non-generic contains the contract with the View and the definition of the handlers.
  * The derived class is generic and contains the implementation of the contract defined by the base class.
* At runtime we update the page application model of each page to update the model type with the implementation class.

### CODE REVIEW

**Comment by javiercn** [comment]
This is a WIP. I'm waiting on a change on MVC to propagate, but I've tried this out by hijacking my nuget package cache and it works without issues for all the views that I've hit.

It needs cleanup, but I've put the PR here so that we can see how it works.

File: samples/IdentitySample.DefaultUI/Data/Migrations/20180124052307_ApplicationUser.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
```
Comment (javiercn) [inline samples/IdentitySample.DefaultUI/Data/Migrations/20180124052307_ApplicationUser.cs]: I'll regenerate the migrations and remove all this stuff

File: src/UI/Areas/Identity/Pages/Account/ExternalLogin.cshtml.cs [hunk 2]
```diff
@@ -43,20 +31,57 @@ public class InputModel
             public string Email { get; set; }
         }
 
-        public IActionResult OnGetAsync()
+        public virtual IActionResult OnGet()
```
Comment (javiercn) [inline src/UI/Areas/Identity/Pages/Account/ExternalLogin.cshtml.cs]: Compact this using expression methods

**Comment by HaoK** [comment]
Can you give me an example of what this actually enables?  Is this all just to allow them to specify their own TUser?

**Comment by HaoK** [comment]
Or is this a step in making it easier to replace a particular view/page?  If that's possible now, can you demonstrate that in the Default UI sample?

**Comment by HaoK** [comment]
The main point I'm getting at is that adding the Name/Age fields to the user is only half of it, they'd most likely want to add support for these fields in the register/manage pages...

If this now enabled the ability to just drop in a custom register/manage and entering/editing the age, that would be pretty compelling.

**Comment by javiercn** [comment]
@haok Yes to both things. This change allows the user to change TUser while reusing all the existing UI.

Overriding views was enabled through precompilation. One example of an overriden view is /Areas/Identity/Pages/_ViewStart.cshtml that is overriden on the sample to make the Identity UI use the main app layout.

I’ll add a sample overriding the register page too if you like. I didn’t do it yet cause I wanted to verify that it worked with all our existing pages.

**Comment by HaoK** [comment]
Ah cool, yeah then can we add some more overriden views/flows to the sample demonstrating this...

Maybe just let them enter their Full name during registration and then let them edit that in Manage.  With out an example its hard for me to see what all the other changes are for, but I'm assuming the IdentityDefaultUI attribute + type is related to overriding.

So to summarize this change, it looks like all of the new abstract base PageModels are basically are the equivalent of the interfaces for each page.  Why are OnGet/OnPost implemented to throw instead of just abstract?

File: src/UI/Areas/Identity/Services/ExternalLoginsServices.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Identity.UI.Services
+{
+    public abstract class ExternalLoginsServices
```
Comment (HaoK) [inline src/UI/Areas/Identity/Services/ExternalLoginsServices.cs]: This seems kinda yucky to have to do this, basically wrapping a call to signInManager in an abstract class

File: src/UI/Areas/Identity/Pages/Account/ForgotPassword.cshtml.cs [hunk 4]
```diff
@@ -31,7 +24,24 @@ public class InputModel
...
         }
 
-        public async Task<IActionResult> OnPostAsync()
+        public virtual Task<IActionResult> OnPostAsync()
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    internal class ForgotPasswordModel<TUser> : ForgotPasswordModel where TUser : IdentityUser
```
Comment (HaoK) [inline src/UI/Areas/Identity/Pages/Account/ForgotPassword.cshtml.cs]: So constraining TUser to be IdentityUser is kinda bad.  Identity is supposed to work against any TUser.  Our templates used to be completely agnostic and only go through the userManager to access data from TUser, only recently was this relaxed a bit to directly access things like user.Id.  

Once we start doing things like this, the ui will basically be even more tightly coupled to the default pocos provided in Identity.Stores.

That might be fine for the default UI, but just one of the original goals of identity was give apps as much control over the user poco as possible.

Basically stuff like this won't work...

```
services.AddIdentity<CustomNotAnIdentityUser, CustomNotIdentityRole>().
   .AddCustomMyOwnStore()
   .AddDefaultUI()
```

Comment (javiercn) [reply src/UI/Areas/Identity/Pages/Account/ForgotPassword.cshtml.cs]: Absolutely agree. I will file a bug for us to revisit this in preview2

File: src/UI/Areas/Identity/Services/ExternalLoginsServices.cs (hunk 3 above)
Comment (javiercn) [reply src/UI/Areas/Identity/Services/ExternalLoginsServices.cs]: Yeah, It’s a limitation due to the lack of open genetics. I’m gonna try and figure out a better solution. Moving forward avoid injecting anything depending on TUser into the view

File: src/UI/IdentityBuilderUIExtensions.cs [hunk 5]
```diff
@@ -33,8 +33,13 @@ public static IdentityBuilder AddDefaultUI(this IdentityBuilder builder)
         {
             AddAdditionalApplicationParts(builder);
 
-            builder.Services.ConfigureOptions<IdentityDefaultUIConfigureOptions>();
+            builder.Services.ConfigureOptions(
+                typeof(IdentityDefaultUIConfigureOptions<>)
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/UI/IdentityBuilderUIExtensions.cs]: There is. I close on the user type here and I can just rely on it in the configuration class. This is the pattern Identity follows for services that depend on the user.

[the remark this answers is not in the document]
Comment (HaoK) [reply src/UI/IdentityBuilderUIExtensions.cs]: I'm with @pranavkm seems easier to just pass in UserType and that lets you get rid some of the TUser : IdentityUser constraints as well

[the remark this answers is not in the document]
Comment (javiercn) [reply src/UI/IdentityBuilderUIExtensions.cs]: @HaoK @pranavkm That will not work. I need to pass in the type (I can't create an instance) because IdentityDefaultUIConfigureOptions has dependencies from DI (IHostingEnvironment).

Passing the user type is more involved/impossible as it would require to pass in a factory method and resolve within that method the hosting environment and pass in the user type.

[the remark this answers is not in the document]
Comment (HaoK) [reply src/UI/IdentityBuilderUIExtensions.cs]: So my concern mostly is that this looks a bit like generic hell where we have a whole lot of complexity just trying to tunnel the TUser to the right place...  This now going to show up in the scaffolded code too?

[the remark this answers is not in the document]
Comment (HaoK) [reply src/UI/IdentityBuilderUIExtensions.cs]: Basically I think the whole public non generic model that throws not implemented + internal derived generic model is kinda gross

[the remark this answers is not in the document]
Comment (javiercn) [reply src/UI/IdentityBuilderUIExtensions.cs]: No

[the remark this answers is not in the document]
Comment (HaoK) [reply src/UI/IdentityBuilderUIExtensions.cs]: Okay, if this weird model stuff isn't going to show up in the scaffolded code, its not as much of an issue

**Comment by javiercn** [comment]
The PR got broken on the rebase because it was pointing to 2.1

**Comment by javiercn** [comment]
185482c8cfdcf4a267fb8b90e3fce48a00ab3076
